x-pack/plugins/alerting/server/saved_objects/migrations.test.ts

```diff
@@ -1012,6 +1012,424 @@ describe('successful migrations', () => {
       });
     });
   });
+
+  describe('7.15.0', () => {
+    test('security solution is migrated to saved object references if it has 1 exceptionsList', () => {
+      const migration7150 = getMigrations(encryptedSavedObjectsSetup)['7.15.0'];
+      const alert = getMockData({
+        alertTypeId: 'siem.signals',
+        params: {
+          note: 'some note', // extra data to ensure we do not overwrite other params
+          exceptionsList: [
+            {
+              id: '123',
+              list_id: '456',
```

Comment:
Don't know much about exception lists so this may be a silly question. What is the `list_id`? Is it another SO ID? Also curious why `namespace_type: single` is stored here? Will these be updated when all the `single` SO types are updated to `multiple-isolated` to make SOs share capable?

Reply:
>What is the list_id? Is it another SO ID?

In one case it is. In another case it is not. In a real world example like below with our `siem.signals`:

query:
```json
GET .kibana/_search
{
  "query": {
    "terms": {
      "alert.alertTypeId": [
        "siem.signals"
      ]
    }
  },
  "size": 10000
}
```

Response (truncated):
```json
              "exceptionsList" : [
                {
                  "list_id" : "endpoint_list",
                  "namespace_type" : "agnostic",
                  "id" : "endpoint_list",
                  "type" : "endpoint"
                },
                {
                  "list_id" : "cd152d0d-3590-4a45-a478-eed04da7936b",
                  "namespace_type" : "single",
                  "id" : "50e3bd70-ef1b-11eb-ad71-7de7959be71c",
                  "type" : "detection"
                }
              ],
```

If it has `"list_id": "endpoint_list"` and `"namespace_type": "agnostic"` then it is a relationship to an agnostic SO with an SO id of `endpoint_list` with the saved object type of, `exception-list-agnostic`. If it has a `list_id` of anything else such as a UUID and a `"namespace_type: "single"` it has a relationship to a non-agnostic saved object of the type `exception-list`.  

When written we did not have the ability to share a mapping in SO's between agnostic and non-agnostic types. I don't know if we can today or not, but the main point is we have two SO types with two mappings that are identical called `exception-list-agnostic` and `exception-list` so that we can create a list that is either agnostic and global vs. being local to a space.

You can see both of these SO types like so:
```json
GET .kibana-hassanabad19/_search
{
  "query": {
    "terms": {
      "type": [
        "exception-list-agnostic",
        "exception-list"
      ]
    }
  }
}
```

>Also curious why namespace_type: single is stored here?

We use it for queries to determine if we are to query against the agnostic SO mapping/type vs. the non SO mapping/type. 

>Will these be updated when all the single SO types are updated to multiple-isolated to make SOs share capable?

No, it will not. If we don't need to duplicate another mapping for a `multiple-isolated` type and use the same mapping/type when we have a type of `single` I don't think it would matter that much. If we need to distinguish between the types and the SO type can change, then we will probably only rely on this value if it is set to `agnostic` and assume otherwise we have to do additional querying to get if it is `single` vs `multipile-isolated`.

We do have a bug here and in an earlier PR I made (part 1 PR), where I am not checking against the word `agnostic` and then adjusting and changing the SO type from `exception-list` to `exception-list-agnostic`

Reply:
@FrankHassanabad Thanks for the explanation! I'm just trying to make sure we're not leaving any dangling SO IDs that could potentially get re-generated in 8.0, and I'm still a little confused. In this example,

```
[{
	"list_id": "cd152d0d-3590-4a45-a478-eed04da7936b",
	"namespace_type": "single",
	"id": "50e3bd70-ef1b-11eb-ad71-7de7959be71c",
	"type": "detection"
}]
```

the exception list saved object ID is `50e3bd70-ef1b-11eb-ad71-7de7959be71c` and that is pulled into the references array. And it says `namespace_type: "single"` because the `exception-list` SO is `single`? If that's correct, that looks good/

What is `cd152d0d-3590-4a45-a478-eed04da7936b`? Is that another exception list saved object ID? A single namespace type exception list SO ID? If it is, this could also be regenerated in 8.0?

Reply:
>the exception list saved object ID is 50e3bd70-ef1b-11eb-ad71-7de7959be71c and that is pulled into the references array. And it says namespace_type: "single" because the exception-list SO is single?

Correct

>What is cd152d0d-3590-4a45-a478-eed04da7936b?  Is that another exception list saved object ID?

No, that is not a saved object ID. That is a regular UUID the code generates within the REST API that we use as the identifier of the list as well as the 1 to M relationship between lists and list items since we could have large volumes of list items.

For example to get all list containers we can do the following with a `list_id`:

```json
# Gets all the list containers attributes such as name, descripion, tags, etc... 
GET .kibana/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "exception-list.list_id": [
              "cd152d0d-3590-4a45-a478-eed04da7936b"
            ]
          }
        },
        {
          "terms": {
            "exception-list.list_type": [
              "list"
            ]
          }
        }
      ]
    }
  }
}
```

This returns (trunacated):

```json
        "_source" : {
          "exception-list" : {
            "created_at" : "2021-07-27T20:43:36.519Z",
            "created_by" : "yo",
            "description" : "kql test rule 1",
            "immutable" : false,
            "list_id" : "cd152d0d-3590-4a45-a478-eed04da7936b", <--- List id here
            "list_type" : "list", <-- Our type is that of "list" and we should only get 1 document back
            "name" : "kql test rule 1",
            "os_types" : [ ],
            "tags" : [ ],
            "tie_breaker_id" : "82b5199f-d409-43fc-a867-1dd62e748c9e",
            "type" : "detection",
            "updated_by" : "yo",
            "version" : 1
          },
          "type" : "exception-list",
          "references" : [ ],
          "migrationVersion" : {
            "exception-list" : "7.12.0"
          },
          "coreMigrationVersion" : "8.0.0",
          "updated_at" : "2021-07-27T20:43:36.718Z"
        }
```

To get all of that list container's items we can do the following:

```json
# Gets all the list items of a particular list container such as a list item's name, description, tags, etc...
GET .kibana/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "exception-list.list_id": [
              "cd152d0d-3590-4a45-a478-eed04da7936b" <-- Find all list items belonging to this list
            ]
          }
        },
        {
          "terms": {
            "exception-list.list_type": [
              "item" <-- We want type item
            ]
          }
        }
      ]
    }
  }
}
```

This returns (truncated):

```json
        "_source" : {
          "exception-list" : {
            "comments" : [
              {
                "comment" : "Test Exception 1",
                "created_at" : "2021-07-27T20:43:57.100Z",
                "created_by" : "yo",
                "id" : "68fd5c83-ee15-485b-a50a-0bdbeefeaaf8"
              }
            ],
            "created_at" : "2021-07-27T20:43:57.100Z",
            "created_by" : "yo",
            "description" : "kql test rule 1 - exception list item",
            "entries" : [
              {
                "field" : "host.name",
                "operator" : "included",
                "type" : "match",
                "value" : "kibana00.siem.estc.dev"
              }
            ],
            "item_id" : "9023a3de-4305-4f4e-b493-9769b0be5533", <-- Our list item id which is also not a SO id
            "list_id" : "cd152d0d-3590-4a45-a478-eed04da7936b", <-- The list we belong to
            "list_type" : "item", <-- We are a list item type
            "name" : "kql test rule 1 - exception list item",
            "os_types" : [ ],
            "tags" : [ ],
            "tie_breaker_id" : "2f8137ef-df3e-47ba-9e91-dea746de2fdc",
            "type" : "simple",
            "updated_by" : "yo"
          },
          "type" : "exception-list",
          "references" : [ ],
          "migrationVersion" : {
            "exception-list" : "7.12.0"
          },
          "coreMigrationVersion" : "8.0.0",
          "updated_at" : "2021-07-27T20:43:57.220Z"
        }
```